File: ControlLab/src/main/java/org/marsroboticsassociation/controllab/arm/RigidArmPlant.java
package org.marsroboticsassociation.controllab.arm;

import org.marsroboticsassociation.controllib.sim.ArmMotorSim;
import org.marsroboticsassociation.controllib.sim.EncoderSim;

/**
 * {@link ArmPlant} backed by the rigid single-inertia {@link ArmMotorSim}. The motor and load are
 * one body, so the motor-side link tracks the load exactly and the teeth are always "engaged".
 */
class RigidArmPlant implements ArmPlant {

    private final ArmPlantConfig cfg;
    private ArmMotorSim sim;

    RigidArmPlant(ArmPlantConfig cfg, double initialAngleRad) {
        this.cfg = cfg;
        rebuild(initialAngleRad);
    }

    private void rebuild(double initialAngleRad) {
        sim = new ArmMotorSim(cfg.kS, cfg.kG, cfg.kV, cfg.kA,
                cfg.ticksPerRev, cfg.gearRatio,
                cfg.encoderZeroOffsetRad, cfg.minAngleRad, cfg.maxAngleRad,
                initialAngleRad);
        sim.setDisturbanceVoltage(cfg.disturbanceVoltage);
        applyEncoder();
    }

    private void applyEncoder() {
        switch (cfg.encoderKind) {
            case CONTROL_HUB:
                sim.setEncoder(EncoderSim.controlHub(cfg.encoderSeed));
                break;
            case EXPANSION_HUB:
                sim.setEncoder(EncoderSim.expansionHub(cfg.encoderSeed));
                break;
            case NONE:
            default:
                sim.setEncoder(new EncoderSim());
                break;
        }
    }

    @Override public void step(double dt, double power, double hubVoltage) {
        sim.step(dt, power, hubVoltage);
    }

    @Override public int getPositionTicks()            { return sim.getPositionTicks(); }
    @Override public double getVelocityTps()           { return sim.getVelocityTps(); }
    @Override public double getTruePositionRad()       { return sim.getTruePositionRad(); }
    @Override public double getTrueVelocityRadPerSec() { return sim.getTrueVelocityRadPerSec(); }
    // Rigid: the motor side is the load side.
    @Override public double getMotorPositionRad()      { return sim.getTruePositionRad(); }
    @Override public boolean isEngaged()               { return true; }
    @Override public double getBacklashRad()           { return 0.0; }

    @Override public void seedFrom(double loadRad, double loadVel) {
        rebuild(loadRad);
    }

    @Override public void applyLiveParams() {
        // Only the disturbance is settable in place on the rigid sim.
        sim.setDisturbanceVoltage(cfg.disturbanceVoltage);
    }
}

File: ControlLab/src/main/kotlin/org/marsroboticsassociation/controllab/arm/RigidArmPlant.kt
package org.marsroboticsassociation.controllab.arm

import org.marsroboticsassociation.controllib.sim.ArmMotorSim
import org.marsroboticsassociation.controllib.sim.EncoderSim

/**
 * [ArmPlant] backed by the rigid single-inertia [ArmMotorSim]. The motor and load are one body, so
 * the motor-side link tracks the load exactly and the teeth are always "engaged".
 */
internal class RigidArmPlant(
    private val cfg: ArmPlantConfig,
    initialAngleRad: Double,
) : ArmPlant {
    private lateinit var sim: ArmMotorSim

    init {
        rebuild(initialAngleRad)
    }

    private fun rebuild(initialAngleRad: Double) {
        sim =
            ArmMotorSim(
                cfg.kS,
                cfg.kG,
                cfg.kV,
                cfg.kA,
                cfg.ticksPerRev,
                cfg.gearRatio,
                cfg.encoderZeroOffsetRad,
                cfg.minAngleRad,
                cfg.maxAngleRad,
                initialAngleRad,
            )
        sim.setDisturbanceVoltage(cfg.disturbanceVoltage)
        applyEncoder()
    }

    private fun applyEncoder() {
        when (cfg.encoderKind) {
            ArmPlantConfig.EncoderKind.CONTROL_HUB ->
                sim.setEncoder(EncoderSim.controlHub(cfg.encoderSeed))
            ArmPlantConfig.EncoderKind.EXPANSION_HUB ->
                sim.setEncoder(EncoderSim.expansionHub(cfg.encoderSeed))
            ArmPlantConfig.EncoderKind.NONE -> sim.setEncoder(EncoderSim())
        }
    }

    override fun step(dt: Double, power: Double, hubVoltage: Double) {
        sim.step(dt, power, hubVoltage)
    }

    override val positionTicks: Int
        get() = sim.getPositionTicks()

    override val velocityTps: Double
        get() = sim.getVelocityTps()

    override val truePositionRad: Double
        get() = sim.getTruePositionRad()

    override val trueVelocityRadPerSec: Double
        get() = sim.getTrueVelocityRadPerSec()

    // Rigid: the motor side is the load side.
    override val motorPositionRad: Double
        get() = sim.getTruePositionRad()

    override val isEngaged: Boolean
        get() = true

    override val backlashRad: Double
        get() = 0.0

    override fun restComplianceRadPerVolt(): Double = 0.0

    override fun seedFrom(loadRad: Double, loadVel: Double) {
        rebuild(loadRad)
    }

    override fun applyLiveParams() {
        // Only the disturbance is settable in place on the rigid sim.
        sim.setDisturbanceVoltage(cfg.disturbanceVoltage)
    }
}
